# 4. Machine learning model

---

### 4.0. Libraries and dataset

In [15]:
import pandas as pd
import sqlite3 as sq
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import utils as eda

# Model libraries
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score
from sklearn.linear_model import Lasso
from sklearn.linear_model import Ridge

train_data = pd.read_csv(r'C:\Users\Francesc\Documents\GitHub\Regularized-linear-regression\data\processed\clean_dataset_train.csv')
test_data = pd.read_csv(r'C:\Users\Francesc\Documents\GitHub\Regularized-linear-regression\data\processed\clean_dataset_test.csv')
target = 'Obesity_number'

X_train = train_data.drop([target], axis = 1)
y_train = train_data[target]
X_test = test_data.drop([target], axis = 1)
y_test = test_data[target]

## 4.1. Linear regression model

### 4.1.1. Initialization and training of the model

In [16]:
model = LinearRegression()
model.fit(X_train, y_train)

LinearRegression()

### 4.1.2. Model prediction

In [17]:
# Model prediction
y_pred = model.predict(X_test)
# model scores
linear_mse = round(mean_squared_error(y_test, y_pred),2)
linear_rmse = round(np.sqrt(linear_mse),2)
linear_r2 = round(r2_score(y_test, y_pred),4)
# Print variables
print(f"Mean Squared Error (MSE): {linear_mse}")
print(f"Root Mean Squared Error (RMSE): {linear_rmse}")
print(f"Coefficient of determination (R^2): {linear_r2}")

Mean Squared Error (MSE): 47684572.67
Root Mean Squared Error (RMSE): 6905.4
Coefficient of determination (R^2): 0.9782


## 4.2. Lasso model

In [18]:
lasso_model = Lasso(alpha = 0.1, max_iter = 300)
lasso_model.fit(X_train, y_train)

score = lasso_model.score(X_test, y_test)
print("Coefficients:", lasso_model.coef_)
print("R2 score:", round(score,4))


Coefficients: [ 71446.62208915  12979.59434516 -17864.39627483  -4471.74656277
  17503.27895054]
R2 score: 0.9784


c:\Users\Francesc\AppData\Local\Programs\Python\Python312\Lib\site-packages\sklearn\linear_model\_coordinate_descent.py:628: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duality gap: 1.044e+11, tolerance: 1.602e+09
  model = cd_fast.enet_coordinate_descent(


## 4.3. Ridge model

In [19]:
ridge_model = Ridge(alpha = 0.1, max_iter = 300)
ridge_model.fit(X_train, y_train)

score = ridge_model.score(X_test, y_test)
print("Coefficients:", ridge_model.coef_)
print("R2 score:", round(score,4))



Coefficients: [ 72928.84213576  13282.65850633 -18630.79193788  -4610.34346328
  16627.99362689]
R2 score: 0.9782


---

# Conclusions

After performing the ML analysis with 5 predictors, the following R2 scores were achieved:

|--| LINEAR | LASSO | RIDGE |
|--|--------|-------|-------|
|R2| 0.9782 | 0.9784| 0.9782|

It is possible to see that there is no difference between Linear and Ridge, and Lasso model has hardly improved the model by two decimal. 

The final score for Lasso linear regression is 0,9784.

